Anuclim and Updated Tox build configuration

tlogan2000 opened the pull request:
<!--Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #411
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (minor / major / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** <!--(Bug fix, feature, docs update, etc.)-->
Adds functions for a subset anuclim bioclimatic indicators  (those based on temp and precipitation

* **Does this PR introduce a breaking change?** <!--(Has there been an API change?)-->
No

* **Other information**:
Docs strings have been added but we could possibly add an automodule .rst similar to ICCLIM?

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+    r""" ANUCLIM Isothermality
+    The mean diurnal range divided by the temperature range.
+
+    Parameters
+    ----------
+    tasmin : xarray.DataArray
+      average daily minimum temperature [℃] or [K] at daily, weekly, or monthly frequency
+
+    tasmax : xarray.DataArray
+      average daily maximum temperature [℃] or [K] at daily, weekly, or monthly frequency
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    Parameters
    ----------
    tasmin : xarray.DataArray
      average daily minimum temperature [℃] or [K] at daily, weekly, or monthly frequency
    tasmax : xarray.DataArray
      average daily maximum temperature [℃] or [K] at daily, weekly, or monthly frequency
    freq : str
      Resampling frequency; Defaults to "YS".
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+    r""" ANUCLIM Precipitation Seasonality (C of V)
+    The annual precipitation Coefficient of Variation (C of V) expressed in percent. Calculated as the standard deviation
+    of precipitation values for a given year expressed as a percentage of the mean of those values.
+    See : https://fennerschool.anu.edu.au/files/anuclim61.pdf ch. 6
```
Wouldn't it be better to put outside references under `Notes`?

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+
+    Parameters
+    ----------
+    tas : xarray.DataArray
+      Mean temperature [℃] or [K] at daily, weekly, or monthly frequency
+
+    op : str
+        Operation to perform :  'warmest' calculate warmest quarter ; 'coldest' calculate coldest quarter
+
+    input_freq : str
+        Input data time frequency - One of 'daily', 'weekly' or 'monthly'
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    tas : xarray.DataArray
      Mean temperature [℃] or [K] at daily, weekly, or monthly frequency.
    op : str
      Operation to perform :  'warmest' calculate warmest quarter ; 'coldest' calculate coldest quarter.
    input_freq : str
      Input data time frequency - One of 'daily', 'weekly' or 'monthly'.
    freq : str
      Resampling frequency; Defaults to "YS".
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+
+    """
+    # determine input data frequency
+    # determine input data frequency
```
```suggestion
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+    pr: xarray.DataArray,
+    op: str = None,
+    input_freq: str = None,
+    freq="YS",
```
```suggestion
    freq: str = "YS",
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+
+    Parameters
+    ----------
+    tas : xarray.DataArray
+      Mean temperature [℃] or [K] at daily, weekly, or monthly frequency
+
+    pr : xarray.DataArray
+      Total precipitation rate at daily, weekly, or monthly frequency
+
+    op : str
+        Operation to perform :  'wettest' calculate for the wettest quarter ; 'dryest' calculate for the dryest quarter
+
+    input_freq : str
+        Input data time frequency - One of 'daily', 'weekly' or 'monthly'
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    tas : xarray.DataArray
      Mean temperature [℃] or [K] at daily, weekly, or monthly frequency.
    pr : xarray.DataArray
      Total precipitation rate at daily, weekly, or monthly frequency.
    op : str
      Operation to perform :  'wettest' calculate for the wettest quarter ; 'dryest' calculate for the dryest quarter.
    input_freq : str
      Input data time frequency - One of 'daily', 'weekly' or 'monthly'.

    freq : str
      Resampling frequency; Defaults to "YS".
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+
+    Parameters
+    ----------
+    pr : xarray.DataArray
+      Total precipitation rate at daily, weekly, or monthly frequency
+
+    op : str
+        Operation to perform :  'wettest' calculate wettest quarter ; 'dryest' calculate dryest quarter
+
+    input_freq : str
+        Input data time frequency - One of 'daily', 'weekly' or 'monthly'
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    pr : xarray.DataArray
      Total precipitation rate at daily, weekly, or monthly frequency.
    op : str
      Operation to perform :  'wettest' calculate wettest quarter ; 'dryest' calculate dryest quarter.
    input_freq : str
      Input data time frequency - One of 'daily', 'weekly' or 'monthly'.
    freq : str
      Resampling frequency; Defaults to "YS".
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+    -------
+    xarray.DataArray
+       Total precipitation values of the wettest/dryest quarter of each year.
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+
+    Parameters
+    ----------
+    pr : xarray.DataArray
+      Total precipitation rate at daily, weekly, or monthly frequency
+
+    tas : xarray.DataArray
+      Mean temperature [℃] or [K] at daily, weekly, or monthly frequency
+
+    op : str
+        Operation to perform: 'warmest' calculate for the warmest quarter ; 'coldest' calculate for the coldest quarter
+
+    input_freq : str
+        Input data time frequency - One of 'daily', 'weekly' or 'monthly'
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    pr : xarray.DataArray
      Total precipitation rate at daily, weekly, or monthly frequency.
    tas : xarray.DataArray
      Mean temperature [℃] or [K] at daily, weekly, or monthly frequency.
    op : str
      Operation to perform: 'warmest' calculate for the warmest quarter ; 'coldest' calculate for the coldest quarter.
    input_freq : str
      Input data time frequency - One of 'daily', 'weekly' or 'monthly'.

    freq : str
      Resampling frequency; Defaults to "YS".
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+def prcptot(pr: xarray.DataArray, input_freq: str = None, freq: str = "YS"):
+    r"""ANUCLIM Accumulated total precipitation.
+
+    Parameters
+    ----------
+    pr : xarray.DataArray
+        Total precipitation flux [mm d-1], [mm week-1], [mm month-1] or similar
+
+    input_freq : str
+        Input data time frequency - One of 'daily', 'weekly' or 'monthly'
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    Parameters
    ----------
    pr : xarray.DataArray
      Total precipitation flux [mm d-1], [mm week-1], [mm month-1] or similar.
    input_freq : str
      Input data time frequency - One of 'daily', 'weekly' or 'monthly'.
    freq : str
      Resampling frequency; Defaults to "YS".
```

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+
+    Parameters
+    ----------
+    pr : xarray.DataArray
+        Total precipitation flux [mm d-1], [mm week-1], [mm month-1] or similar
+
+    op : str
+        Operation to perform :  'wettest' calculate wettest period ; 'dryest' calculate dryest period
+
+    input_freq : str
+        Input data time frequency - One of 'daily', 'weekly' or 'monthly'
+
+    freq : str
+      Resampling frequency; Defaults to "YS".
```
```suggestion
    pr : xarray.DataArray
      Total precipitation flux [mm d-1], [mm week-1], [mm month-1] or similar.
    op : str
      Operation to perform :  'wettest' calculate wettest period ; 'dryest' calculate dryest period.
    input_freq : str
      Input data time frequency - One of 'daily', 'weekly' or 'monthly'.
    freq : str
      Resampling frequency; Defaults to "YS".
```

huard commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,624 @@
+    Returns
+    -------
+    xarray.DataArray
+      The isothermailty value expressed as a percent.
```
```suggestion
      The isothermality value expressed as a percent.
```

aulemahal commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,599 @@
+
+
+@declare_units("mm", pr="[precipitation]")
+def prcptot(pr: xarray.DataArray, input_freq: str = None, freq: str = "YS"):
```
So this does the same as `precip_accumulation` but with the `input_freq` flag added? I'm not sure it's a good idea to add  it to the `xclim.indices` namespace...  We could maybe add the `input_freq` management to the existing  precip_accumulation`? Or rename this to remove confusion between the two? Or forbid `input_freq="daily"`?

tlogan2000 replied:
I agree not too sure what to do here really.... I originally wanted to just do a wrapped partial of precip_accum in the anuclim module but the return applies a ```pint_multiply(pr, 1 * units.day, "mm")```  so if input data is weekly for example it ends up dividing by 7  ...   I thought it was weird to add support for other frequencies to a single index of our 'standard' library (i.e. non anuplin indices)..  `

I can remove daily support but would there a way for me to wrap two separate calls into the "P12_AnnualPrecip" function in the mapped module..  right now it does this ``"P12_AnnualPrecip": wrapped_partial(indices.prcptot, freq="YS")``   would need a lambda if else kind of thing??

aulemahal replied:
Maybe you could move the input_freq-dependent part of the code to the `__build_anuclim` function, so its available to the Anuclim module but never made public?

tlogan2000 replied:
> Maybe you could move the input_freq-dependent part of the code to the `__build_anuclim` function, so its available to the Anuclim module but never made public?

Euhh, Might need some help there ...  :/   Sounds good but not sure how to implement... 

Would this mean the _anucplin.prcptot() would be removed? (Great if possible) and we could somehow shoehorn precip_accum to do what we want ?

aulemahal commented:
1) There was a bug in `_get_from_other_extreme` that I fixed. Also, added tests using dask to `TestTempWetDryPrecipWarmColdQuarter` . However, we still have problems : when `input_freq='daily'` and the data is stored in a dask array, the preliminary resampling operation will create small chunks along 'time' and the `resample` will fail as the chunks are smaller than half the window size... I skipped those tests for now, as the user can easily do the resampling and rechunk before.

2) Shouldn't we use 'D', 'W' and 'M' for `input_freq`? One day, we might use a `xr.infer_freq` to fill this field automatically and that function will return one of those. And we could test with : `input_freq.startswith('D')`.

Zeitsperre commented on xclim/indices/_anuclim.py:
```diff
@@ -0,0 +1,599 @@
+        return pr.resample(time=freq).max(dim="time", keep_attrs=True)
+    elif op == "dryest":
+        return pr.resample(time=freq).min(dim="time", keep_attrs=True)
+    else:
+        raise NotImplementedError(
+            f'Unknown operation "{op}" ; op parameter but be one of "wettest" or "dryest"'
+        )
```
```suggestion
    raise NotImplementedError(
        f'Unknown operation "{op}" ; op parameter but be one of "wettest" or "dryest"'
    )
```

Zeitsperre commented:
>     2. Shouldn't we use 'D', 'W' and 'M' for `input_freq`? One day, we might use a `xr.infer_freq` to fill this field automatically and that function will return one of those. And we could test with : `input_freq.startswith('D')`.

It would be good if we followed xarray/pandas-like time_freq conventions as closely as possible. I wonder if there are any helper functions from those libraries that we can lean on...?

tlogan2000 commented:
> > 1. Shouldn't we use 'D', 'W' and 'M' for `input_freq`? One day, we might use a `xr.infer_freq` to fill this field automatically and that function will return one of those. And we could test with : `input_freq.startswith('D')`.
> 
> It would be good if we followed xarray/pandas-like time_freq conventions as closely as possible. I wonder if there are any helper functions from those libraries that we can lean on...?

ok. good call.  I think eventually we would like to see a version where input_freq isn't event there and we get the frequency via xr.infer_freq.  However, it currently doesn't support non-standard calendars so this was a work-around

Zeitsperre commented:
@tlogan2000 Would you be upset if I did a find and replace for `dryest` to `driest` ?

tlogan2000 commented:
> @tlogan2000 Would you be upset if I did a find and replace for `dryest` to `driest` ?

haha. ok with me.  I had actually coded driest everywhere while @aulemahal had 'dryest'.  I went with 'dryest' as I was not sure what non-native english speakers would be expecting and personally it was  either/or

Zeitsperre commented:
I had the impression that `dryest` was an out-west cowboy colloquial spelling, haha.

tlogan2000 commented:
> I had the impression that `dryest` was an out-west cowboy colloquial spelling, haha.

Websters doesn't seem to have 'dryest' so I guess driest would be the most correct

tlogan2000 commented:
> 1. There was a bug in `_get_from_other_extreme` that I fixed, my bad. Also, added tests using dask to `TestTempWetDryPrecipWarmColdQuarter` . However, we still have problems : when `input_freq='daily'` and the data is stored in a dask array, the preliminary resampling operation will create small chunks along 'time' and the `resample` will fail as the chunks are smaller than half the window size... I skipped those tests for now, as the user can easily do the resampling and rechunk before.


Would this be a problem for all the 'wetdry'/'warmcold' quarter indices?  i.e. anything with a rolling or is it specific to those using the sub function?

aulemahal commented:
> Would this be a problem for all the 'wetdry'/'warmcold' quarter indices? i.e. anything with a rolling or is it specific to those using the sub function?

It's specific to the indices resampling daily to weekly and then calling a `xr.apply_ufunc`, so right now only the cross-indices (tas of driest, pr of warmest and so on). 

> out-west cowboy colloquial spelling
Sadly, it was only keb engrish.

huard commented:
In `tg_mean_warmcold_quarter`, if `input_freq` is daily, why not set `wind` to 13*7 ?

huard commented:
Also, the dosctring should probably mention that the `freq` should include an offset to avoid chopping the warmest quarter in 2 (e.g. AS-JUL) if you're in the SH.

tlogan2000 commented:
> Also, the dosctring should probably mention that the `freq` should include an offset to avoid chopping the warmest quarter in 2 (e.g. AS-JUL) if you're in the SH.

Right.  This will affect my mapping...  I mapped most as partial with freq='YS'.  ..  Might therfore need to make sure that input freq is annual for some indices (seasonality at the very least)

tlogan2000 commented:
> In `tg_mean_warmcold_quarter`, if `input_freq` is daily, why not set `wind` to 13*7 ?

This would be fine as well.  Initial thought was to try reproduce anuclim output as closely as possible but maybe we don't want to force ourselves to accept their 'weekly' input limitation

huard commented:
So I wrote a `_quarter` function that could be reused in a couple of indices. I tested it on two of them, I'll let you decide if you want to use that or not. If we decide to change the behavior, then there is only one place it would change.

huard commented:
I think you could simplify the logic by using "min", "max" as the `op`.

aulemahal commented:
Woups! We implemented different methods at the same time. But they were easily merged. I moved the "from_other" lazy indexing machinery to `xclim.indices.run_length.lazy_indexing` that can now accept ND inputs. (and to do so, changed `_np_argops` to `_xr_argops`.  For reference, the new way was 1.5x faster on my machine. 
Also, propagated your fix (@huard ) to `prcptot_warmcold_quarter`.

huard commented:
@aulemahal Cyclical import issue possibly due to: 
```
xclim/sdba/__init__.py:1:0: R0401: Cyclic import (xclim -> xclim.core.options -> xclim.locales -> xclim.core.formatting) (cyclic-import)
```
